src/libtiled/tilelayer.cpp

```diff
@@ -312,12 +312,24 @@ void TileLayer::rotate(RotateDirection direction)
     mGrid = newGrid;
 }
 
-void TileLayer::rotateHexagonal(RotateDirection direction)
+void TileLayer::rotateHexagonal(RotateDirection direction, Map::StaggerIndex staggerIndex, Map::StaggerAxis staggerAxis)
 {
-    int newWidth = mHeight;
-    int newHeight = mWidth;
+    Hex botRight(mWidth, mHeight, staggerIndex, staggerAxis);
+    Hex topRight(mWidth, 0, staggerIndex, staggerAxis);
+    Hex center(mWidth / 2, mHeight / 2, staggerIndex, staggerAxis);
+
+    botRight -= center;
+    topRight -= center;
+
+    botRight.rotate(RotateRight);
+    topRight.rotate(RotateRight);
+
+    int newWidth = topRight.toStaggered(staggerIndex, staggerAxis).x() * 2 + 2;
+    int newHeight = botRight.toStaggered(staggerIndex, staggerAxis).y() * 2 + 2;
```

Comment:
If a rectangle is rotated 60 degrees clockwise around the center, then the top right corner will be the furthest right point, the bottom left will be the furthest left by the same amount, and top left will be the highest, bottom right the lowest. Similar logic applies for anti-clockwise rotation, and the sizes will be the same, so one direction can be chosen, regardless of intended rotation direction. So we take the two corners, transform them around the center, rotate, and take the respective largest axis. This value is of the center, so it is doubled to get the maximum new width/height. The two is added because of potential integer rounding in regards to the center. 

Of course a problem is that sometimes the corners will be empty, so the enlargement is unnecessary. I don't believe there is an easy way to account for this, thus I enlarge, then trim it down later.

In short: This process produces the smallest rectangle which contains the original rotated by 60 degrees.